### PR TITLE
- Deinline destructor.

### PR DESCRIPTION
**Pull request by baldersheim**
- Move instead of copy = > noexcept move construtors.
- make_unique/make_shared.
- c++11 for loops

@toregge or @vekterli PR

### CODE REVIEW

**Review by toregge — APPROVED** [review]

**Review by vekterli — COMMENTED** [review]
:+1: I think a lot of the added explicit `std::move`s could be elided with the addition of some overloads with rvalue reference qualifiers (see comment)

File: messagebus/src/tests/routingcontext/routingcontext.cpp [hunk 1]
```diff
@@ -73,21 +73,19 @@ CustomPolicy::CustomPolicy(CustomPolicyFactory &factory) :
...
 
     const std::vector<Route> &all = ctx.getAllRecipients();
     if (_factory._expectedAll.size() == all.size()) {
         ctx.trace(1, make_string("Got %d expected recipients.", (uint32_t)all.size()));
-        for (std::vector<Route>::const_iterator it = all.begin();
-             it != all.end(); ++it)
-        {
-            if (find(_factory._expectedAll.begin(), _factory._expectedAll.end(), it->toString()) != _factory._expectedAll.end()) {
-                ctx.trace(1, make_string("Got expected recipient '%s'.", it->toString().c_str()));
+        for (const auto & it : all) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/tests/routingcontext/routingcontext.cpp:82]: route it is

File: documentapi/src/tests/policies/testframe.cpp [hunk 2]
```diff
@@ -111,8 +111,8 @@ TestFrame::testSelect(const std::vector<string> &expected)
     std::vector<mbus::RoutingNode*> selected;
     if (!select(selected, expected.size())) {
         LOG(error, "Failed to select recipients.");
-        for (size_t i = 0; i < selected.size(); ++i) {
-            LOG(error, "Selected: %s", selected[i]->getRoute().toString().c_str());
+        for (auto & i : selected) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply documentapi/src/tests/policies/testframe.cpp:114]: i -> node

File: messagebus/src/tests/advancedrouting/advancedrouting.cpp [hunk 3]
```diff
@@ -123,7 +123,7 @@ Test::testAdvanced(TestData &data)
     data._srcServer.mb.setupRouting(RoutingSpec().addTable(RoutingTableSpec(SimpleProtocol::NAME)
                                                            .addHop(HopSpec("bar", "dst/bar"))
                                                            .addHop(HopSpec("baz", "dst/baz"))
-                                                           .addRoute(RouteSpec("baz").addHop("baz"))));
+                                                           .addRoute(std::move(RouteSpec("baz").addHop("baz")))));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply messagebus/src/tests/advancedrouting/advancedrouting.cpp:126]: Ah, tried, but forgot the & at the end....
Earlier on it was the number and placement of * that was the issue. Now it is even harder with &, && in all possible corners...

**Pull request merged**